earth_enterprise/docker/image-definition/Dockerfile.build-env.rhel-6.template

```diff
@@ -0,0 +1,28 @@
+# We only need a few context files, rather than the whole repository:
+# --[ TST-Dockerfile: context-dir=earth_enterprise/docker/image-definition ]--
+# Prompt the user to enter variables required to build this Dockerfile:
+# --[ TST-Dockerfile: required-vars=REDHAT_USERNAME ]--
+# Avoid displaying the values of some variables:
+# --[ TST-Dockerfile: required-password-vars=REDHAT_PASSWORD ]--
+
+FROM registry.access.redhat.com/rhel6
+
+
+# Copy shell scripts to /root/bin in the Docker container:
+RUN mkdir -p /root/opengee/bin
+COPY scripts /root/opengee/bin
+
+# Make a build directory:
+RUN mkdir -p /root/opengee/build
+
+
+# Install the build environment.
+#   We register the Docker image with the Red Hat repository provider (you
+# need to have a username and password) so we can install Yum packages.  We
+# unregister when we finish, so we're not taking up a machine license.
+WORKDIR /root/opengee/build
+RUN subscription-manager register \
+        --username=${REDHAT_USERNAME_ESCAPED} --password=${REDHAT_PASSWORD_ESCAPED} \
```

Comment:
If there's any way possible, we should really try to avoid printing out the password for this line

Reply:
Yes.  What kinds of scenarios should we protect against?

I was thinking of using image variables, at first, but the Docker that comes with Ubuntu 14 is too old, and doesn't support them, so I'm expanding variables in Dockerfile templates instead.  The variables a container is built with a recorded, so, if we did that, we'd have to delete them from the final container, and delete all intermediate container layers, and terminated images.

Currently, the expanded template is stored in a temporary Docker file that is deleted after the build.  Also, a docker image may have the Dockerfile it was built from left behind.  (I need to check for this on RHEL.)

Here are some options:

1. Put the password in a file inside the container.  Delete the file after the password is consumed.

This may be a simple way to avoid having the password in the Docker command logs (which appear on the screen).  There are a lot of scenarios that doesn't protect a user from, but, at least, it won't be shown on the screen.

2. Put the password in an encrypted file in the container.  Delete the file after the password is consumed.

2.1. We could use a temporary 1-time encryption cipher that is automatically generated for each build.

This way we'd passing around the cipher, which wouldn't be useful after the password is consumed and the password-containing file deleted.  The advantage is that we won't be storing the password on disk, so, even if the deleted disk blocks become exposed later, the original password would still not be exposed.  (This may be a threat scenario that requires an attacker stealing someone's disk and running an analysis on it.)

2.2. If the user has multiple passwords in a vault. We could copy the vault, and pass the vault password instead of the RHEL subscription password.

That doesn't improve security from the approach above, but might more convenient in some cases.

3. We could obtain passwords in the container over a(n encrypted) network connection to the host OS.

Passwords would not appear in command logs, and won't be stored on disk.

A tool like [Vault](https://github.com/hashicorp/vault) can encrypt passwords in a file, or provide them over a network connection.

In all of these cases the password would pass in clear text through memory.

Reply:
Sure, it may not be the most secure ever, but it's gotta be better than displaying the password in plain text on the screen :) I was very surprised to see that.

Reply:
It seems that using a named pipe works fine.  That way it doesn't reach the disk.  I just need to finish the scripting.